File: rb/lib/twitter-text/extractor.rb
# Copyright 2018 Twitter, Inc.
# Licensed under the Apache License, Version 2.0
# http://www.apache.org/licenses/LICENSE-2.0

# encoding: utf-8
require 'idn'

class String
  # Helper function to count the character length by first converting to an
  # array.  This is needed because with unicode strings, the return value
  # of length may be incorrect
  def codepoint_length
    if respond_to? :codepoints
      length
    else
      chars.kind_of?(Enumerable) ? chars.to_a.size : chars.size
    end
  end

  # Helper function to convert this string into an array of unicode code points.
  def to_codepoint_a
    @to_codepoint_a ||= if chars.kind_of?(Enumerable)
      chars.to_a
    else
      codepoint_array = []
      0.upto(codepoint_length - 1) { |i| codepoint_array << [chars.slice(i)].pack('U') }
      codepoint_array
    end
  end
end

# Helper functions to return code point offsets instead of byte offsets.
class MatchData
  def char_begin(n)
    if string.respond_to? :codepoints
      self.begin(n)
    else
      string[0, self.begin(n)].codepoint_length
    end
  end

  def char_end(n)
    if string.respond_to? :codepoints
      self.end(n)
    else
      string[0, self.end(n)].codepoint_length
    end
  end
end

module Twitter
  module TwitterText
    # A module for including Tweet parsing in a class. This module provides function for the extraction and processing
    # of usernames, lists, URLs and hashtags.
    module Extractor extend self

      # Maximum URL length as defined by Twitter's backend.
      MAX_URL_LENGTH = 4096

      # The maximum t.co path length that the Twitter backend supports.
      MAX_TCO_SLUG_LENGTH = 40

      URL_PROTOCOL_LENGTH = "https://".length

      # Remove overlapping entities.
      # This returns a new array with no overlapping entities.
      def remove_overlapping_entities(entities)
        # sort by start index
        entities = entities.sort_by{|entity| entity[:indices].first}

        # remove duplicates
        prev = nil
        entities.reject!{|entity| (prev && prev[:indices].last > entity[:indices].first) || (prev = entity) && false}
        entities
      end

      # Extracts all usernames, lists, hashtags and URLs  in the Tweet <tt>text</tt>
      # along with the indices for where the entity ocurred
      # If the <tt>text</tt> is <tt>nil</tt> or contains no entity an empty array
      # will be returned.
      #
      # If a block is given then it will be called for each entity.
      def extract_entities_with_indices(text, options = {}, &block)
        config = options[:config] || Twitter::TwitterText::Configuration.default_configuration

        # extract all entities
        entities = extract_urls_with_indices(text, options) +
                   extract_hashtags_with_indices(text, :check_url_overlap => false) +
                   extract_mentions_or_lists_with_indices(text) +
                   extract_cashtags_with_indices(text)
        entities += extract_emoji_with_indices(text) if config.emoji_parsing_enabled

        return [] if entities.empty?

        entities = remove_overlapping_entities(entities)

        entities.each(&block) if block_given?
        entities
      end

      # Extracts a list of all usernames mentioned in the Tweet <tt>text</tt>. If the
      # <tt>text</tt> is <tt>nil</tt> or contains no username mentions an empty array
      # will be returned.
      #
      # If a block is given then it will be called for each username.
      def extract_mentioned_screen_names(text, &block) # :yields: username
        screen_names = extract_mentioned_screen_names_with_indices(text).map{|m| m[:screen_name]}
        screen_names.each(&block) if block_given?
        screen_names
      end

      # Extracts a list of all usernames mentioned in the Tweet <tt>text</tt>
      # along with the indices for where the mention ocurred.  If the
      # <tt>text</tt> is nil or contains no username mentions, an empty array
      # will be returned.
      #
      # If a block is given, then it will be called with each username, the start
      # index, and the end index in the <tt>text</tt>.
      def extract_mentioned_screen_names_with_indices(text) # :yields: username, start, end
        return [] unless text

        possible_screen_names = []
        extract_mentions_or_lists_with_indices(text) do |screen_name, list_slug, start_position, end_position|
          next unless list_slug.empty?
          possible_screen_names << {
            :screen_name => screen_name,
            :indices => [start_position, end_position]
          }
        end

        if block_given?
          possible_screen_names.each do |mention|
            yield mention[:screen_name], mention[:indices].first, mention[:indices].last
          end
        end

        possible_screen_names
      end

      # Extracts a list of all usernames or lists mentioned in the Tweet <tt>text</tt>
      # along with the indices for where the mention ocurred.  If the
      # <tt>text</tt> is nil or contains no username or list mentions, an empty array
      # will be returned.
      #
      # If a block is given, then it will be called with each username, list slug, the start
      # index, and the end index in the <tt>text</tt>. The list_slug will be an empty stirng
      # if this is a username mention.
      def extract_mentions_or_lists_with_indices(text) # :yields: username, list_slug, start, end
        return [] unless text =~ /[@＠]/

        possible_entries = []
        text.to_s.scan(Twitter::TwitterText::Regex[:valid_mention_or_list]) do |before, at, screen_name, list_slug|
          match_data = $~
                        after = $'
          unless after =~ Twitter::TwitterText::Regex[:end_mention_match]
            start_position = match_data.char_begin(3) - 1
            end_position = match_data.char_end(list_slug.nil? ? 3 : 4)
            possible_entries << {
              :screen_name => screen_name,
              :list_slug => list_slug || "",
              :indices => [start_position, end_position]
            }
          end
        end

        if block_given?
          possible_entries.each do |mention|
            yield mention[:screen_name], mention[:list_slug], mention[:indices].first, mention[:indices].last
          end
        end

        possible_entries
      end

      # Extracts the username username replied to in the Tweet <tt>text</tt>. If the
      # <tt>text</tt> is <tt>nil</tt> or is not a reply nil will be returned.
      #
      # If a block is given then it will be called with the username replied to (if any)
      def extract_reply_screen_name(text) # :yields: username
        return nil unless text

        possible_screen_name = text.match(Twitter::TwitterText::Regex[:valid_reply])
        return unless possible_screen_name.respond_to?(:captures)
        return if $' =~ Twitter::TwitterText::Regex[:end_mention_match]
        screen_name = possible_screen_name.captures.first
        yield screen_name if block_given?
        screen_name
      end

      # Extracts a list of all URLs included in the Tweet <tt>text</tt>. If the
      # <tt>text</tt> is <tt>nil</tt> or contains no URLs an empty array
      # will be returned.
      #
      # If a block is given then it will be called for each URL.
      def extract_urls(text, &block) # :yields: url
        urls = extract_urls_with_indices(text).map{|u| u[:url]}
        urls.each(&block) if block_given?
        urls
      end

      # Extracts a list of all URLs included in the Tweet <tt>text</tt> along
      # with the indices. If the <tt>text</tt> is <tt>nil</tt> or contains no
      # URLs an empty array will be returned.
      #
      # If a block is given then it will be called for each URL.
      def extract_urls_with_indices(text, options = {:extract_url_without_protocol => true}) # :yields: url, start, end
        return [] unless text && (options[:extract_url_without_protocol] ? text.index(".") : text.index(":"))
        urls = []

        text.to_s.scan(Twitter::TwitterText::Regex[:valid_url]) do |all, before, url, protocol, domain, port, path, query|
          valid_url_match_data = $~

                                  start_position = valid_url_match_data.char_begin(3)
          end_position = valid_url_match_data.char_end(3)

          # If protocol is missing and domain contains non-ASCII characters,
          # extract ASCII-only domains.
          if !protocol
            next if !options[:extract_url_without_protocol] || before =~ Twitter::TwitterText::Regex[:invalid_url_without_protocol_preceding_chars]
            last_url = nil
            domain.scan(Twitter::TwitterText::Regex[:valid_ascii_domain]) do |ascii_domain|
              next unless is_valid_domain(url.length, ascii_domain, protocol)
              last_url = {
                :url => ascii_domain,
                :indices => [start_position + $~.char_begin(0),
                             start_position + $~.char_end(0)]
              }
              urls << last_url
            end

            # no ASCII-only domain found. Skip the entire URL
            next unless last_url

            # last_url only contains domain. Need to add path and query if they exist.
            if path
              # last_url was not added. Add it to urls here.
              last_url[:url] = url.sub(domain, last_url[:url])
              last_url[:indices][1] = end_position
            end
          else
            # In the case of t.co URLs, don't allow additional path characters
            if url =~ Twitter::TwitterText::Regex[:valid_tco_url]
              next if $1 && $1.length > MAX_TCO_SLUG_LENGTH
              url = $&
                    end_position = start_position + url.codepoint_length
            end

            next unless is_valid_domain(url.length, domain, protocol)

            urls << {
              :url => url,
              :indices => [start_position, end_position]
            }
          end
        end
        urls.each{|url| yield url[:url], url[:indices].first, url[:indices].last} if block_given?
        urls
      end

      # Extracts a list of all hashtags included in the Tweet <tt>text</tt>. If the
      # <tt>text</tt> is <tt>nil</tt> or contains no hashtags an empty array
      # will be returned. The array returned will not include the leading <tt>#</tt>
      # character.
      #
      # If a block is given then it will be called for each hashtag.
      def extract_hashtags(text, &block) # :yields: hashtag_text
        hashtags = extract_hashtags_with_indices(text).map{|h| h[:hashtag]}
        hashtags.each(&block) if block_given?
        hashtags
      end

      # Extracts a list of all hashtags included in the Tweet <tt>text</tt>. If the
      # <tt>text</tt> is <tt>nil</tt> or contains no hashtags an empty array
      # will be returned. The array returned will not include the leading <tt>#</tt>
      # character.
      #
      # If a block is given then it will be called for each hashtag.
      def extract_hashtags_with_indices(text, options = {:check_url_overlap => true}) # :yields: hashtag_text, start, end
        return [] unless text =~ /[#＃]/

        tags = []
        text.scan(Twitter::TwitterText::Regex[:valid_hashtag]) do |before, hash, hash_text|
          match_data = $~
                        start_position = match_data.char_begin(2)
          end_position = match_data.char_end(3)
          after = $'
          unless after =~ Twitter::TwitterText::Regex[:end_hashtag_match]
            tags << {
              :hashtag => hash_text,
              :indices => [start_position, end_position]
            }
          end
        end

        if options[:check_url_overlap]
          # extract URLs
          urls = extract_urls_with_indices(text)
          unless urls.empty?
            tags.concat(urls)
            # remove duplicates
            tags = remove_overlapping_entities(tags)
            # remove URL entities
            tags.reject!{|entity| !entity[:hashtag] }
          end
        end

        tags.each{|tag| yield tag[:hashtag], tag[:indices].first, tag[:indices].last} if block_given?
        tags
      end

      # Extracts a list of all cashtags included in the Tweet <tt>text</tt>. If the
      # <tt>text</tt> is <tt>nil</tt> or contains no cashtags an empty array
      # will be returned. The array returned will not include the leading <tt>$</tt>
      # character.
      #
      # If a block is given then it will be called for each cashtag.
      def extract_cashtags(text, &block) # :yields: cashtag_text
        cashtags = extract_cashtags_with_indices(text).map{|h| h[:cashtag]}
        cashtags.each(&block) if block_given?
        cashtags
      end

      # Extracts a list of all cashtags included in the Tweet <tt>text</tt>. If the
      # <tt>text</tt> is <tt>nil</tt> or contains no cashtags an empty array
      # will be returned. The array returned will not include the leading <tt>$</tt>
      # character.
      #
      # If a block is given then it will be called for each cashtag.
      def extract_cashtags_with_indices(text) # :yields: cashtag_text, start, end
        return [] unless text =~ /\$/

        tags = []
        text.scan(Twitter::TwitterText::Regex[:valid_cashtag]) do |before, dollar, cash_text|
          match_data = $~
                        start_position = match_data.char_begin(2)
          end_position = match_data.char_end(3)
          tags << {
            :cashtag => cash_text,
            :indices => [start_position, end_position]
          }
        end

        tags.each{|tag| yield tag[:cashtag], tag[:indices].first, tag[:indices].last} if block_given?
        tags
      end

      def extract_emoji_with_indices(text) # :yields: emoji, start, end
        emoji = []
        text.scan(Twitter::TwitterText::Regex[:valid_emoji]) do |emoji_text|
          match_data = $~
                        start_position = match_data.char_begin(0)
          end_position = match_data.char_end(0)
          emoji << {
            :emoji => emoji_text,
            :indices => [start_position, end_position]
          }
        end
        emoji
      end

      def is_valid_emoji(text)
        begin
          raise ArgumentError.new("invalid empty emoji") unless text
          entities = extract_emoji_with_indices(text)
          entities.count == 1 && entities[0][:emoji] == text
        rescue Exception
          # On error don't consider this a valid domain.
          return false
        end
      end

      def is_valid_domain(url_length, domain, protocol)
        begin
          raise ArgumentError.new("invalid empty domain") unless domain
          original_domain_length = domain.length
          encoded_domain = IDN::Idna.toASCII(domain)
          updated_domain_length = encoded_domain.length
          url_length += (updated_domain_length - original_domain_length) if (updated_domain_length > original_domain_length)
          url_length += URL_PROTOCOL_LENGTH unless protocol
          url_length <= MAX_URL_LENGTH
        rescue Exception
          # On error don't consider this a valid domain.
          return false
        end
      end
    end
  end
end

File: java/src/main/java/com/twitter/twittertext/Extractor.java
// Copyright 2018 Twitter, Inc.
// Licensed under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0

package com.twitter.twittertext;

import java.net.IDN;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;
import java.util.regex.Matcher;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;

/**
 * A class to extract usernames, lists, hashtags and URLs from Tweet text.
 */
public class Extractor {
  /**
   * The maximum url length that the Twitter backend supports.
   */
  public static final int MAX_URL_LENGTH = 4096;

  /**
   * The maximum t.co path length that the Twitter backend supports.
   */
  public static final int MAX_TCO_SLUG_LENGTH = 40;

  /**
   * The backend adds http:// for normal links and https to *.twitter.com URLs
   * (it also rewrites http to https for URLs matching *.twitter.com).
   * We're better off adding https:// all the time. By making the assumption that
   * URL_GROUP_PROTOCOL_LENGTH is https, the trade off is we'll disallow a http URL
   * that is 4096 characters.
   */
  private static final int URL_GROUP_PROTOCOL_LENGTH = "https://".length();

  public static class Entity {
    public enum Type {
      URL, HASHTAG, MENTION, CASHTAG
    }

    protected int start;
    protected int end;
    protected final String value;
    // listSlug is used to store the list portion of @mention/list.
    protected final String listSlug;
    protected final Type type;

    protected String displayURL = null;
    protected String expandedURL = null;

    public Entity(int start, int end, String value, String listSlug, Type type) {
      this.start = start;
      this.end = end;
      this.value = value;
      this.listSlug = listSlug;
      this.type = type;
    }

    public Entity(int start, int end, String value, Type type) {
      this(start, end, value, null, type);
    }

    public Entity(Matcher matcher, Type type, int groupNumber) {
      // Offset -1 on start index to include @, # symbols for mentions and hashtags
      this(matcher, type, groupNumber, -1);
    }

    public Entity(Matcher matcher, Type type, int groupNumber, int startOffset) {
      this(matcher.start(groupNumber) + startOffset, matcher.end(groupNumber),
          matcher.group(groupNumber), type);
    }

    @Override
    public boolean equals(Object obj) {
      if (this == obj) {
        return true;
      }

      if (!(obj instanceof Entity)) {
        return false;
      }

      Entity other = (Entity) obj;

      return this.type.equals(other.type) &&
          this.start == other.start &&
          this.end == other.end &&
          this.value.equals(other.value);
    }

    @Override
    public int hashCode() {
      return this.type.hashCode() + this.value.hashCode() + this.start + this.end;
    }

    @Override
    public String toString() {
      return value + "(" + type + ") [" + start + "," + end + "]";
    }

    public Integer getStart() {
      return start;
    }

    public Integer getEnd() {
      return end;
    }

    public String getValue() {
      return value;
    }

    public String getListSlug() {
      return listSlug;
    }

    public Type getType() {
      return type;
    }

    public String getDisplayURL() {
      return displayURL;
    }

    public void setDisplayURL(String displayURL) {
      this.displayURL = displayURL;
    }

    public String getExpandedURL() {
      return expandedURL;
    }

    public void setExpandedURL(String expandedURL) {
      this.expandedURL = expandedURL;
    }
  }

  protected boolean extractURLWithoutProtocol = true;

  /**
   * Create a new extractor.
   */
  public Extractor() {
  }

  private void removeOverlappingEntities(List<Entity> entities) {
    // sort by index
    Collections.sort(entities, new Comparator<Entity>() {
      public int compare(Entity e1, Entity e2) {
        return e1.start - e2.start;
      }
    });

    // Remove overlapping entities.
    // Two entities overlap only when one is URL and the other is hashtag/mention
    // which is a part of the URL. When it happens, we choose URL over hashtag/mention
    // by selecting the one with smaller start index.
    if (!entities.isEmpty()) {
      Iterator<Entity> it = entities.iterator();
      Entity prev = it.next();
      while (it.hasNext()) {
        Entity cur = it.next();
        if (prev.getEnd() > cur.getStart()) {
          it.remove();
        } else {
          prev = cur;
        }
      }
    }
  }

  /**
   * Extract URLs, @mentions, lists and #hashtag from a given text/tweet.
   *
   * @param text text of tweet
   * @return list of extracted entities
   */
  public List<Entity> extractEntitiesWithIndices(String text) {
    List<Entity> entities = new ArrayList<Entity>();
    entities.addAll(extractURLsWithIndices(text));
    entities.addAll(extractHashtagsWithIndices(text, false));
    entities.addAll(extractMentionsOrListsWithIndices(text));
    entities.addAll(extractCashtagsWithIndices(text));

    removeOverlappingEntities(entities);
    return entities;
  }

  /**
   * Extract @username references from Tweet text. A mention is an occurrence of @username anywhere
   * in a Tweet.
   *
   * @param text of the tweet from which to extract usernames
   * @return List of usernames referenced (without the leading @ sign)
   */
  public List<String> extractMentionedScreennames(String text) {
    if (isEmptyString(text)) {
      return Collections.emptyList();
    }

    List<String> extracted = new ArrayList<String>();
    for (Entity entity : extractMentionedScreennamesWithIndices(text)) {
      extracted.add(entity.value);
    }
    return extracted;
  }

  /**
   * Extract @username references from Tweet text. A mention is an occurrence of @username anywhere
   * in a Tweet.
   *
   * @param text of the tweet from which to extract usernames
   * @return List of usernames referenced (without the leading @ sign)
   */
  public List<Entity> extractMentionedScreennamesWithIndices(String text) {
    List<Entity> extracted = new ArrayList<Entity>();
    for (Entity entity : extractMentionsOrListsWithIndices(text)) {
      if (entity.listSlug == null) {
        extracted.add(entity);
      }
    }
    return extracted;
  }

  /**
   * Extract @username and an optional list reference from Tweet text. A mention is an occurrence
   * of @username anywhere in a Tweet. A mention with a list is a @username/list.
   *
   * @param text of the tweet from which to extract usernames
   * @return List of usernames (without the leading @ sign) and an optional lists referenced
   */
  public List<Entity> extractMentionsOrListsWithIndices(String text) {
    if (isEmptyString(text)) {
      return Collections.emptyList();
    }

    // Performance optimization.
    // If text doesn't contain @/＠ at all, the text doesn't
    // contain @mention. So we can simply return an empty list.
    boolean found = false;
    for (char c : text.toCharArray()) {
      if (c == '@' || c == '＠') {
        found = true;
        break;
      }
    }
    if (!found) {
      return Collections.emptyList();
    }

    List<Entity> extracted = new ArrayList<Entity>();
    Matcher matcher = Regex.VALID_MENTION_OR_LIST.matcher(text);
    while (matcher.find()) {
      String after = text.substring(matcher.end());
      if (!Regex.INVALID_MENTION_MATCH_END.matcher(after).find()) {
        if (matcher.group(Regex.VALID_MENTION_OR_LIST_GROUP_LIST) == null) {
          extracted.add(new Entity(matcher, Entity.Type.MENTION,
              Regex.VALID_MENTION_OR_LIST_GROUP_USERNAME));
        } else {
          extracted.add(new Entity(matcher.start(Regex.VALID_MENTION_OR_LIST_GROUP_USERNAME) - 1,
              matcher.end(Regex.VALID_MENTION_OR_LIST_GROUP_LIST),
              matcher.group(Regex.VALID_MENTION_OR_LIST_GROUP_USERNAME),
              matcher.group(Regex.VALID_MENTION_OR_LIST_GROUP_LIST),
              Entity.Type.MENTION));
        }
      }
    }
    return extracted;
  }

  /**
   * Extract a @username reference from the beginning of Tweet text. A reply is an occurrence
   * of @username at the beginning of a Tweet, preceded by 0 or more spaces.
   *
   * @param text of the tweet from which to extract the replied to username
   * @return username referenced, if any (without the leading @ sign).
   * Returns null if this is not a reply.
   */
  public String extractReplyScreenname(String text) {
    if (text == null) {
      return null;
    }

    Matcher matcher = Regex.VALID_REPLY.matcher(text);
    if (matcher.find()) {
      String after = text.substring(matcher.end());
      if (Regex.INVALID_MENTION_MATCH_END.matcher(after).find()) {
        return null;
      } else {
        return matcher.group(Regex.VALID_REPLY_GROUP_USERNAME);
      }
    } else {
      return null;
    }
  }

  /**
   * Extract URL references from Tweet text.
   *
   * @param text of the tweet from which to extract URLs
   * @return List of URLs referenced.
   */
  @Nonnull
  public List<String> extractURLs(@Nullable String text) {
    if (isEmptyString(text)) {
      return Collections.emptyList();
    }

    final List<String> urls = new ArrayList<>();
    for (Entity entity : extractURLsWithIndices(text)) {
      urls.add(entity.value);
    }
    return urls;
  }

  /**
   * Extract URL references from Tweet text.
   *
   * @param text of the tweet from which to extract URLs
   * @return List of URLs referenced.
   */
  @Nonnull
  public List<Entity> extractURLsWithIndices(@Nullable String text) {
    if (isEmptyString(text) ||
        (extractURLWithoutProtocol ? text.indexOf('.') : text.indexOf(':')) == -1) {
      // Performance optimization.
      // If text doesn't contain '.' or ':' at all, text doesn't contain URL,
      // so we can simply return an empty list.
      return Collections.emptyList();
    }

    final List<Entity> urls = new ArrayList<>();

    final Matcher matcher = Regex.VALID_URL.matcher(text);
    while (matcher.find()) {
      final String protocol = matcher.group(Regex.VALID_URL_GROUP_PROTOCOL);
      if (isEmptyString(protocol)) {
        // skip if protocol is not present and 'extractURLWithoutProtocol' is false
        // or URL is preceded by invalid character.
        if (!extractURLWithoutProtocol
            || Regex.INVALID_URL_WITHOUT_PROTOCOL_MATCH_BEGIN
            .matcher(matcher.group(Regex.VALID_URL_GROUP_BEFORE)).matches()) {
          continue;
        }
      }
      String url = matcher.group(Regex.VALID_URL_GROUP_URL);
      int start = matcher.start(Regex.VALID_URL_GROUP_URL);
      int end = matcher.end(Regex.VALID_URL_GROUP_URL);
      final Matcher tcoMatcher = Regex.VALID_TCO_URL.matcher(url);
      if (tcoMatcher.find()) {
        final String tcoUrl = tcoMatcher.group(0);
        final String tcoUrlSlug = tcoMatcher.group(1);
        // In the case of t.co URLs, don't allow additional path characters and
        // ensure that the slug is under 40 chars.
        if (tcoUrlSlug.length() > MAX_TCO_SLUG_LENGTH) {
          continue;
        } else {
          url = tcoUrl;
          end = start + url.length();
        }
      }
      final String host = matcher.group(Regex.VALID_URL_GROUP_DOMAIN);
      if (isValidHostAndLength(url.length(), protocol, host)) {
        urls.add(new Entity(start, end, url, Entity.Type.URL));
      }
    }

    return urls;
  }

  /**
   * Verifies that the host name adheres to RFC 3490 and 1035
   * Also, verifies that the entire url (including protocol) doesn't exceed MAX_URL_LENGTH
   *
   * @param originalUrlLength The length of the entire URL, including protocol if any
   * @param protocol The protocol used
   * @param originalHost The hostname to check validity of
   * @return true if the host is valid
   */
  public static boolean isValidHostAndLength(int originalUrlLength, @Nullable String protocol,
                                             @Nullable String originalHost) {
    if (isEmptyString(originalHost)) {
      return false;
    }
    final int originalHostLength = originalHost.length();
    String host;
    try {
      // Use IDN for all host names, if the host is all ASCII, it returns unchanged.
      // It comes with an added benefit of checking host length to be between 1 and 63 characters.
      host = IDN.toASCII(originalHost, IDN.ALLOW_UNASSIGNED);
      // toASCII can throw IndexOutOfBoundsException when the domain name is longer than
      // 256 characters, instead of the documented IllegalArgumentException.
    } catch (IllegalArgumentException | IndexOutOfBoundsException e) {
      return false;
    }
    final int punycodeEncodedHostLength = host.length();
    if (punycodeEncodedHostLength == 0) {
      return false;
    }
    // The punycodeEncoded host length might be different now, offset that length from the URL.
    final int urlLength = originalUrlLength + punycodeEncodedHostLength - originalHostLength;
    // Add the protocol to our length check, if there isn't one,
    // to ensure it doesn't go over the limit.
    final int urlLengthWithProtocol =
        urlLength + (protocol == null ? URL_GROUP_PROTOCOL_LENGTH : 0);
    return urlLengthWithProtocol <= MAX_URL_LENGTH;
  }

  /**
   * Extract #hashtag references from Tweet text.
   *
   * @param text of the tweet from which to extract hashtags
   * @return List of hashtags referenced (without the leading # sign)
   */
  public List<String> extractHashtags(String text) {
    if (isEmptyString(text)) {
      return Collections.emptyList();
    }

    List<String> extracted = new ArrayList<String>();
    for (Entity entity : extractHashtagsWithIndices(text)) {
      extracted.add(entity.value);
    }

    return extracted;
  }

  /**
   * Extract #hashtag references from Tweet text.
   *
   * @param text of the tweet from which to extract hashtags
   * @return List of hashtags referenced (without the leading # sign)
   */
  public List<Entity> extractHashtagsWithIndices(String text) {
    return extractHashtagsWithIndices(text, true);
  }

  /**
   * Extract #hashtag references from Tweet text.
   *
   * @param text of the tweet from which to extract hashtags
   * @param checkUrlOverlap if true, check if extracted hashtags overlap URLs and
   * remove overlapping ones
   * @return List of hashtags referenced (without the leading # sign)
   */
  private List<Entity> extractHashtagsWithIndices(String text, boolean checkUrlOverlap) {
    if (isEmptyString(text)) {
      return Collections.emptyList();
    }

    // Performance optimization.
    // If text doesn't contain #/＃ at all, text doesn't contain
    // hashtag, so we can simply return an empty list.
    boolean found = false;
    for (char c : text.toCharArray()) {
      if (c == '#' || c == '＃') {
        found = true;
        break;
      }
    }
    if (!found) {
      return Collections.emptyList();
    }

    List<Entity> extracted = new ArrayList<Entity>();
    Matcher matcher = Regex.VALID_HASHTAG.matcher(text);

    while (matcher.find()) {
      String after = text.substring(matcher.end());
      if (!Regex.INVALID_HASHTAG_MATCH_END.matcher(after).find()) {
        extracted.add(new Entity(matcher, Entity.Type.HASHTAG, Regex.VALID_HASHTAG_GROUP_TAG));
      }
    }

    if (checkUrlOverlap) {
      // extract URLs
      List<Entity> urls = extractURLsWithIndices(text);
      if (!urls.isEmpty()) {
        extracted.addAll(urls);
        // remove overlap
        removeOverlappingEntities(extracted);
        // remove URL entities
        Iterator<Entity> it = extracted.iterator();
        while (it.hasNext()) {
          Entity entity = it.next();
          if (entity.getType() != Entity.Type.HASHTAG) {
            it.remove();
          }
        }
      }
    }

    return extracted;
  }

  /**
   * Extract $cashtag references from Tweet text.
   *
   * @param text of the tweet from which to extract cashtags
   * @return List of cashtags referenced (without the leading $ sign)
   */
  public List<String> extractCashtags(String text) {
    if (isEmptyString(text)) {
      return Collections.emptyList();
    }

    List<String> extracted = new ArrayList<String>();
    for (Entity entity : extractCashtagsWithIndices(text)) {
      extracted.add(entity.value);
    }

    return extracted;
  }

  /**
   * Extract $cashtag references from Tweet text.
   *
   * @param text of the tweet from which to extract cashtags
   * @return List of cashtags referenced (without the leading $ sign)
   */
  public List<Entity> extractCashtagsWithIndices(String text) {
    if (isEmptyString(text)) {
      return Collections.emptyList();
    }

    // Performance optimization.
    // If text doesn't contain $, text doesn't contain
    // cashtag, so we can simply return an empty list.
    if (text.indexOf('$') == -1) {
      return Collections.emptyList();

    }

    List<Entity> extracted = new ArrayList<Entity>();
    Matcher matcher = Regex.VALID_CASHTAG.matcher(text);

    while (matcher.find()) {
      extracted.add(new Entity(matcher, Entity.Type.CASHTAG, Regex.VALID_CASHTAG_GROUP_CASHTAG));
    }

    return extracted;
  }

  public void setExtractURLWithoutProtocol(boolean extractURLWithoutProtocol) {
    this.extractURLWithoutProtocol = extractURLWithoutProtocol;
  }

  public boolean isExtractURLWithoutProtocol() {
    return extractURLWithoutProtocol;
  }

  /**
   * Modify Unicode-based indices of the entities to UTF-16 based indices.
   * <p>
   * In UTF-16 based indices, Unicode supplementary characters are counted as two characters.
   * <p>
   * This method requires that the list of entities be in ascending order by start index.
   *
   * @param text original text
   * @param entities entities with Unicode based indices
   */
  public void modifyIndicesFromUnicodeToUTF16(String text, List<Entity> entities) {
    IndexConverter convert = new IndexConverter(text);

    for (Entity entity : entities) {
      entity.start = convert.codePointsToCodeUnits(entity.start);
      entity.end = convert.codePointsToCodeUnits(entity.end);
    }
  }

  /**
   * Modify UTF-16-based indices of the entities to Unicode-based indices.
   * <p>
   * In Unicode-based indices, Unicode supplementary characters are counted as single characters.
   * <p>
   * This method requires that the list of entities be in ascending order by start index.
   *
   * @param text original text
   * @param entities entities with UTF-16 based indices
   */
  public void modifyIndicesFromUTF16ToUnicode(String text, List<Entity> entities) {
    IndexConverter convert = new IndexConverter(text);

    for (Entity entity : entities) {
      entity.start = convert.codeUnitsToCodePoints(entity.start);
      entity.end = convert.codeUnitsToCodePoints(entity.end);
    }
  }

  private static boolean isEmptyString(@Nullable CharSequence string) {
    return string == null || string.length() == 0;
  }

  /**
   * An efficient converter of indices between code points and code units.
   */
  private static final class IndexConverter {
    protected final String text;

    // Keep track of a single corresponding pair of code unit and code point
    // offsets so that we can re-use counting work if the next requested
    // entity is near the most recent entity.
    protected int codePointIndex = 0;
    protected int charIndex = 0;

    IndexConverter(String text) {
      this.text = text;
    }

    /**
     * Converts code units to code points
     *
     * @param charIndex Index into the string measured in code units.
     * @return The code point index that corresponds to the specified character index.
     */
    int codeUnitsToCodePoints(int charIndex) {
      if (charIndex < this.charIndex) {
        this.codePointIndex -= text.codePointCount(charIndex, this.charIndex);
      } else {
        this.codePointIndex += text.codePointCount(this.charIndex, charIndex);
      }
      this.charIndex = charIndex;

      // Make sure that charIndex never points to the second code unit of a
      // surrogate pair.
      if (charIndex > 0 && Character.isSupplementaryCodePoint(text.codePointAt(charIndex - 1))) {
        this.charIndex -= 1;
      }
      return this.codePointIndex;
    }

    /**
     * Converts code points to code units
     *
     * @param codePointIndex Index into the string measured in code points.
     * @return the code unit index that corresponds to the specified code point index.
     */
    int codePointsToCodeUnits(int codePointIndex) {
      // Note that offsetByCodePoints accepts negative indices.
      this.charIndex =
          text.offsetByCodePoints(this.charIndex, codePointIndex - this.codePointIndex);
      this.codePointIndex = codePointIndex;
      return this.charIndex;
    }
  }
}
